chain/beacon/sync_manager.go

```diff
@@ -455,14 +466,77 @@ type SyncStream interface {
 	Send(*proto.BeaconPacket) error
 }
 
-// SyncChain holds the receiver logic to reply to a sync request
-func SyncChain(l log.Logger, store CallbackStore, req SyncRequest, stream SyncStream) error {
+// SyncChain holds the receiver logic to reply to a sync request to stream back beacons
+// it stops onces it reaches the end of our current store.
+func SyncChain(l log.Logger, store chain.Store, req SyncRequest, stream SyncStream) error {
+	logger := l.Named("SyncChain")
+
+	fromRound := req.GetFromRound()
+	if fromRound == 0 {
+		return fmt.Errorf("invalid sync request trying to sync from 0")
+	}
+
+	ctx := stream.Context()
+	addr := net.RemoteAddress(ctx)
+	beaconID := beaconIDToSync(l, req, addr)
+
+	last, err := store.Last(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to get last beacon: %w", err)
+	}
+
+	if last.Round < fromRound {
+		return fmt.Errorf("%w %d < %d", chainerrors.ErrNoBeaconStored, last.Round, fromRound)
+	}
+
+	send := func(b *chain.Beacon) error {
+		err := stream.Send(beaconToProto(b, beaconID))
+		if err != nil {
+			logger.Debugw("", "syncer", "streaming_send", "err", err)
+		}
+		return err
+	}
+
+	// we know that last.Round >= fromRound from the above if
+
+	// TODO (dlsniper): During the loop below, we can receive new data
+	//  which may not be observed.
+	//  Investigate if how the storage view updates while the cursor runs.
+
+	// sync up from the store itself
+	err = store.Cursor(ctx, func(ctx context.Context, c chain.Cursor) error {
+		bb, err := c.Seek(ctx, fromRound)
+		for ; bb != nil; bb, err = c.Next(ctx) {
+			// This is needed since send will use a pointer and could result in pointer reassignment
+			bb := bb
+			if err != nil {
+				return err
+			}
+			// Force send the correct
+			if err := send(bb); err != nil {
+				logger.Debugw("Error while sending beacon", "syncer", "cursor_seek")
+				return err
+			}
+		}
+		return err
+	})
+	// We always have ErrNoBeaconStored returned as last value
+	// so let's ignore it and not send it back to the client
+	if errors.Is(err, chainerrors.ErrNoBeaconStored) {
+		return nil
+	}
+	return err
+}
+
+// FollowChain holds the receiver logic to reply to a sync request to stream back beacons. It keeps streaming
+// them using a Callback in the store Put function to do so after reaching the end of the store.
+func FollowChain(l log.Logger, store CallbackStore, req SyncRequest, stream SyncStream) error {
```

Comment:
Im not sure I see where do you fix a bug - can you expand on the PR description ? 
Looking at this PR, it seems you're adding followchain but that doesnt solve the problem of syncing ? The syncing method above seems the same as before ? I'm lost

Reply:
The SyncChain call now always returns instead of hanging forever with a callback in the store on every SyncChain request.
Therefor the stream should now be closed

The FollowChain request is only every used by the `drand sync --follow` command and is meant to follow the chain, but it's canceled when the client cancels its stream. There's no risk that SyncChain doesn't return here after it's done syncing now:
https://github.com/drand/drand/blob/057ab8cd5103ffcf8e6619462ed16f6c89bebaac/core/drand_beacon_public.go#L230

Reply:
ok thx -exactly as you mention in the description, unclear if that really kills the bug, but that will help for sure.
👍